V3 user guide edits (#2129)
GMLC-TDC/HELICS: docs/user-guide/examples/advanced_examples/advanced_orchestration.md

After the two federates pass information between each other -- EV Battery sends SOC, EV Charger instructs whether to keep charging or resample the distributions -- the EV Battery calculates the total power demanded in the last time interval.

## Execution and Results

Execution can be done with either a simple script (provided on the repo), or with Merlin.

### Manual Orchestration Execution

Manual implementation of the co-simulation is done with the helper script `advanced_orchestration.py`, with command line execution:

```

> $ python advanced_orchestration.py

Reviewer: @allisonmcampbell the `advanced_orchestration.py` file doesn't exist in https://github.com/GMLC-TDC/HELICS-Examples/tree/main/user_guide_examples/advanced/advanced_orchestration

Reviewer: I think I figured this one out -- the `master` branch in HELICS-Examples has been renamed to `main`; make sure you update your forks to push changes to `main` instead of `master`. I'm going to merge the changes you made into `main` and do some branch renaming to try to fix the automatic redirects GitHub sets up.

Reviewer: Change is merged -- redirect is working again for links.

Remember to update your fork to use `main` instead of `master`; the instructions GitHub gives for doing this when you visit https://github.com/GMLC-TDC/HELICS-Examples for the first time after it has been renamed are:

```
git branch -m master main
git fetch origin
git branch -u origin/main main
git remote set-head origin -a
```

GitHub should also show this message if you push to the wrong branch: `remote: Heads up! The branch 'master' that you pushed to was renamed to 'main'.`

Reviewer 2: Yes, thank you for the reminder!

In the Merlin spec we will instruct Merlin to execute N number of the
Monte Carlo co-simulations. The number of samples is the number
specified as the `N_SAMPLES` env variable in the env section of the
merlin spec.

```yaml
env:
  variables:
    OUTPUT_PATH: ./UQ_EV_Study
    N_SAMPLES: 10
```

We set the output directory to UQ_EV_Study, this is where all the
output files will be stored. Every co-simulation run executed by
merlin will have it's own subdirectory in `./UQ_EV_Study`.

##### Merlin Step

Remember this step is for Merlin to setup all the files and data it
needs to execute it's jobs. In the Monte Carlo co-simulation there is
a python script we created that will generated the `helics_cli` runner
files that Merlin will use when it executes the `helics_cli`. The
`make_samples_merlin.py` script will also output a csv file that Merlin will
use. The csv file contains all the names of the runner files. Merlin
will go down this list of file names and execute the `helics_cli` for
each file name.

```yaml
merlin:
  samples:
    generate:
      cmd: |
        python3 $(SPECROOT)/make_samples_merlin.py $(N_SAMPLES) $(MERLIN_INFO)
        cp $(SPECROOT)/EVMsgFed.py $(MERLIN_INFO)

>         cp $(SPECROOT)/EVControllerMsgFed.py $(MERLIN_INFO)

Reviewer: @allisonmcampbell I'm also not finding a `make_samples_merlin.py` or the `EVMsgFed.py` files in the HELICS-Examples repository. The closest I'm finding is those two python files and a `make_samples.py` file in https://github.com/GMLC-TDC/HELICS/tree/main/docs/user-guide/examples/user_guide_examples/Example_3

Reviewer 2: `make_samples_merlin.py` now lives here:
HELICS-Examples/user_guide_examples/advanced/advanced_orchestration/simple
renamed EVMsgFed --> Battery
renamed EVControllerMsgFed --> Charger

<!-- identify two options for testing this example:
1. manual
2. Merlin spec for Cosimulation
 -->

# Monte Carlo Co-Simulations

> ```{eval-rst}

Reviewer: This creates an empty  toc (and there is a manually typed toc just below this), so I think the block here can be removed.

  - [Merlin Orchestration Execution](#merlin-orchestration-execution)

## Where is the code?

Code for the Monte Carlo simulation and the
full Merlin spec and be found in the [HELICS Examples Repo](https://github.com/GMLC-TDC/HELICS-Examples/tree/master/user_guide_examples/advanced/advanced_orchestration). If you have issues navigating to the examples, visit the HELICS Gitter page or the user forum on GitHub.

[![](../../../img/advanced_orchestration_github.png)](https://github.com/GMLC-TDC/HELICS-Examples/tree/master/user_guide_examples/advanced/advanced_orchestration)

The necessary files are:

- Python program for Battery federate
- Python program for Charger federate
- Python program to generate `helics_cli` JSON files and execute

## What is this co-simulation doing?

This example walks through how to set up a probabilistic model with Monte Carlo simulations. This Monte Carlo co-simulation is built from a simple two federate example, based on the [Endpoint Federates Example](../fundamental_examples/fundamental_endpoints.md). In this example, there is a Charger federate which publishes voltage and a Battery federate which publishes current.

> All of the HELICS configurations are the same as in the Endpoint example. The internal logic of the federates has been changed for this implementation. The Charger federate assumes the role of _deciding_ if the Battery should continue to charge. The Battery sends a message of its current state of charge (soc, a number between 0 and 1). If the soc is less than 0.9, the Battery is instructed to continue charging, otherwise, it is instructed to cease charging. The Battery federate has all the logic internal for adding energy and selecting a new "battery" (charging rate) if the soc is deemed sufficient. Energy is added to the "battery" according to the previous time interval and the charge rate of the battery. In this way, the only stochastic component to the system is the **selected charge rate**. For example, the Endpoint Example allowed the Battery federate to randomly select batteries of different sizes, and the Charger to select charge rates from a list of options. In this implementation, the battery size (capacity in kWh) is constant.

Reviewer: Reorder `soc` and expanded version, remove the value range from the parenthesis to make it the abbreviation meaning clearer.

```suggestion
All of the HELICS configurations are the same as in the Endpoint example. The internal logic of the federates has been changed for this implementation. The Charger federate assumes the role of _deciding_ if the Battery should continue to charge. The Battery sends a message of its current soc (state of charge), a number between 0 and 1. If the soc is less than 0.9, the Battery is instructed to continue charging, otherwise, it is instructed to cease charging. The Battery federate has all the logic internal for adding energy and selecting a new "battery" (charging rate) if the soc is deemed sufficient. Energy is added to the "battery" according to the previous time interval and the charge rate of the battery. In this way, the only stochastic component to the system is the **selected charge rate**. For example, the Endpoint Example allowed the Battery federate to randomly select batteries of different sizes, and the Charger to select charge rates from a list of options. In this implementation, the battery size (capacity in kWh) is constant.
```

$$ L \sim P(k,\lambda) $$

Let's extend our original assumption that the distribution of chargers is static to Poisson distributed, and let's assume that there are 100 total charging ports:

$$ L1 \sim P(100,0.3) $$
$$ L2 \sim P(100,0.5) $$
$$ L3 \sim P(100,0.2) $$

![](../../../img/EVPoisson.png)

What if we weren't entirely certain that the average values for $L1, L2, L3$ are $0.3, 0.5, 0.2$, we can also sample the averages from a normal distribution centered on these values with reasonable standard deviations. We can say that:

$$ \lambda \sim N(\mu,\sigma) $$

Which means that the input to $L$ is distributed normally with average $\mu$ and standard deviation $\sigma$.

Our final distribution for modeling the anticipated need for each level of charging port in our $k = 100$ EV garage can be written as:

> $$ L \sim P(k,\lambda) $$

Reviewer: I think we need to add a line to the `docs/conf.py` file to configure the MyST-parser to work with dollarmath: https://myst-parser.readthedocs.io/en/latest/syntax/optional.html#math-shortcuts

Author: OK, if you say so. I've never worked with the build process for the docs so somebody else (@phlptp?, @nightlark?) so I propose somebody who knows what they are doing tackle this.

Reviewer: Seems to work now. MyST extensions for math are enabled, and an old conflicting Sphinx markdown table extension has been removed.

Notice that the individual overplotted distributions in the two histograms above are different -- there is more flexibility encoded into the second. The distributions in the second plot describe the following assumptions about the anticipated need for Level 1, 2, and 3 chargers:

1. The number of charging ports is discrete and positive (Poisson).
2. The variance in the number of ports should be normally distributed.
3. The average percentage of Level 1, 2, and 3 chargers is 30%, 50%, and 20%, respectively.
4. The variance around the average is uniformly distributed.
5. The variance is relatively very broad for Level 1, broad for Level 2, and very constrained for Level 3. This means we are very confident in our guess of the average percentage for Level 3 chargers, less confident for Level 2, and much less confident for Level 3.

We have described the individual distributions for each level of charging port. What we don't know, prior to running the Monte Carlo co-simulation, is how these distributions will jointly impact the research question.

#### Research Question Configuration

We want to address the research question: What is the likely power draw that EVs will demand?

> At the beginning of the co-simulation, the distributions defined above will be sampled $N$ times within the EV federate, where $N =$ the number of parking spots/charging ports in the garage. The output of the initial sampling is the number of requested Level 1, 2, and 3 charging ports. The soc for the batteries on board are initialized to a uniform random number between 0.05 and 0.5, and these SOC are sent to the Charger federate. If the SOC of an EV battery is less than 0.9, the EV Controller federate tells the EV battery in the EV federate to continue charging. Otherwise, the EV Charger disconnects the EV battery, and instructs the it to sample a new EV battery from the distributions (one sample).

Reviewer: ```suggestion
At the beginning of the co-simulation, the distributions defined above will be sampled $N$ times within the EV federate, where $N =$ the number of parking spots/charging ports in the garage. The output of the initial sampling is the number of requested Level 1, 2, and 3 charging ports. The SOC for the batteries on board are initialized to a uniform random number between 0.05 and 0.5, and these SOC are sent to the Charger federate. If the SOC of an EV battery is less than 0.9, the EV Controller federate tells the EV battery in the EV federate to continue charging. Otherwise, the EV Charger disconnects the EV battery, and instructs the it to sample a new EV battery from the distributions (one sample).
```

After the two federates pass information between each other -- EV Battery sends SOC, EV Charger instructs whether to keep charging or resample the distributions -- the EV Battery calculates the total power demanded in the last time interval.

## Execution and Results

Execution can be done with either a simple script (provided on the repo), or with Merlin.

### Manual Orchestration Execution

Manual implementation of the co-simulation is done with the helper script `make_samples_manual.py`, with command line execution:

```shell session
$ python make_samples.py
```

This implementation will run a default co-simulation. The default parameters are:

```python

>     samples = 30
>     output_path = os.getcwd()
>     numEVs = 10
>     hours = 24
>     plot = 0
>     run = 1

Reviewer: ```suggestion
samples = 30
output_path = os.getcwd()
numEVs = 10
hours = 24
plot = 0
run = 1
```

```

> This means that we are generating 30 JSON files with unique seeds, we are using the current operating directory as the head for the output path, we are simulating 10 EVs in the co-simulation for one day, we are not running individual plots for each simulation, and we are executing the JSON files with `helics_cli` after they have been created.

Reviewer: ```suggestion
This means that we are generating 30 JSON files with unique seeds, we are using the current operating directory as the head for the output path, we are simulating 10 EVs in the co-simulation for one day, we are not running individual plots for each simulation, and we are running `helics_cli` with the JSON files after they have been created.
```

If we wanted to run a Monte Carlo co-sim with different parameters, this would be:

```shell session
$ python make_samples.py 10 . 100 24*7 0 0
```

> This execution would create 10 JSON files with unique seeds, set the current directory as the head for the output path, simulate 100 EVs for a week, not generate plots with each simulation, and not execution JSONs with `helics_cli`.

Reviewer: ```suggestion
This execution would create 10 JSON files with unique seeds, set the current directory as the head for the output path, simulate 100 EVs for a week, not generate plots with each simulation, and not run `helics-cli` with the generated JSON files.
```

> You may decide to adapt `make_samples_manual.py` to suite your needs within the Merlin environment, in which case you would only need the helper script to create the JSON files. If you elect to execute the JSONs using the helper script, sub directories are created for the `helics_cli` runner JSONs and for the csv results. Results for the default simulation are on the repo and can be used for confirming accurate execution.

Reviewer: ```suggestion
You may decide to adapt `make_samples_manual.py` to suite your needs within the Merlin environment, in which case you would only need the helper script to create the JSON files. If you elect to execute the JSON files using the helper script, sub directories are created for the `helics_cli` runner JSON files and for the csv results. Results for the default simulation are in the repo and can be used for confirming accurate execution.
```

```python

>     out_json = output_path+'/cli_runner_scripts'
>     out_data = output_path+'/results'

Reviewer: ```suggestion
out_json = output_path+'/cli_runner_scripts'
out_data = output_path+'/results'
```

```

> In the runner scripts directory, there will be 30 JSONs. Each will have a unique `seed` parameter, otherwise they will all look identical:

Reviewer: ```suggestion
In the runner scripts directory, there will be 30 JSON files. Each will have a unique `seed` parameter, otherwise they will all look identical:
```

In this specification we will be using the
[helics_cli](https://github.com/GMLC-TDC/helics-cli) to execute each
co-simulation run since this is a Monte Carlo simulation. This means
that helics_cli will be executed multiple times with different
helics_cli runner files.

#### Co-simulation Reproduction

Management of multiple iterations of the co-simulation can be done by setting the seed as a function of the number of brokers, where there will be one broker for each iteration. In Python 3.X:

```python
import argparse

parser = argparse.ArgumentParser(description='EV simulator')
parser.add_argument('--seed', type=int, default=0,
                help='The seed that will be used for our random distribution')
parser.add_argument('--port', type=int, default=-1,
                help='port of the HELICS broker')

args = parser.parse_args()
np.random.seed(args.seed)
```

#### helics_cli in Merlin

Since we are using the helics_cli to manage and execute all the
federates, we need to create these runner files for helics_cli.
There is a provided python script called `make_samples_merlin.py` that will
generate the runner file and a csv file that will be used in the

> study step. Helics_cli will start each of these federates. In the Merlin

Reviewer: ```suggestion
study step. helics_cli will start each of these federates. In the Merlin
```

spec, Merlin will be instructed to execute the helics_cli with all the
generated helics_cli runner files.

#### Merlin Specification

##### Environment

In the Merlin spec we will instruct Merlin to execute N number of the
Monte Carlo co-simulations. The number of samples is the number
specified as the `N_SAMPLES` env variable in the env section of the

> merlin spec.

Reviewer: ```suggestion
Merlin spec.
```

```yaml
env:
  variables:
    OUTPUT_PATH: ./UQ_EV_Study
    N_SAMPLES: 10
```

We set the output directory to UQ_EV_Study, this is where all the
output files will be stored. Every co-simulation run executed by

> merlin will have it's own subdirectory in `./UQ_EV_Study`.

Reviewer: ```suggestion
Merlin will have it's own subdirectory in `./UQ_EV_Study`.
```

```

This tutorial will walk through how to set up a HELICS Monte Carlo simulation using two techniques: (1) in series on a single machine, and (2) in parallel on an HPC cluster using Merlin. We assume that you have already completed the
[**orchestration tutorial with Merlin**](../../advanced_topics/orchestration.md) and have some
familiarity with how Merlin works.

- [Where is the code?](#where-is-the-code)
- [What is this Co-simulation doing?](#what-is-this-co-simulation-doing)
- [Probabilistic Uncertainty Estimation](#probabilistic-uncertainty-estimation)
- [Execution and Results](#execution-and-results)
  - [Manual Orchestration Execution](#manual-orchestration-execution)
  - [Merlin Orchestration Execution](#merlin-orchestration-execution)

## Where is the code?

Code for the Monte Carlo simulation and the

> full Merlin spec and be found in the [HELICS Examples Repo](https://github.com/GMLC-TDC/HELICS-Examples/tree/master/user_guide_examples/advanced/advanced_orchestration). If you have issues navigating to the examples, visit the HELICS Gitter page or the user forum on GitHub.

Author: I don't think there's any reason to be saying this. For users that are new to software development, having frequent reminders pointing them towards help, even for broken links, is a good idea.

> [![](../../../img/advanced_orchestration_github.png)](https://github.com/GMLC-TDC/HELICS-Examples/tree/master/user_guide_examples/advanced/advanced_orchestration)

Author: We've had screenshots for many of the examples but since this one doesn't exist, we'll just remove it for now.

The necessary files are:

- Python program for Battery federate
- Python program for Charger federate

> - Python program to generate `helics_cli` JSON files and execute

Author: Changed.

```json
{
  "federates": [
    {
      "directory": "/Users/camp426/github/HELICS-Examples/user_guide_examples/advanced/advanced_orchestration",
      "exec": "helics_broker --federates=2 --loglevel=data --coretype=tcpss --port 12345",
      "host": "localhost",
      "loglevel": "data",
      "name": "broker_0"
    },
    {
      "directory": "/Users/camp426/github/HELICS-Examples/user_guide_examples/advanced/advanced_orchestration",
      "exec": "python3 Battery.py --port 12345 --seed 10 --numEVs 10 --hours 24 --plot 0 --outdir ~/HELICS-Examples/user_guide_examples/advanced/advanced_orchestration/results",
      "host": "localhost",
      "loglevel": "data",
      "name": "Battery_0"
    },
    {
      "directory": "/Users/camp426/github/HELICS-Examples/user_guide_examples/advanced/advanced_orchestration",
      "exec": "python3 Charger.py --port 12345 --numEVs 10 --hours 24",
      "host": "localhost",
      "loglevel": "data",
      "name": "Charger_0"
    }
  ],
  "name": "Generated by make samples"
}
```

The final result of the default Monte Carlo co-simulation is shown below.

> ![](./MonteCarlo_manual.png)

Author: @allisonmcampbell, can you take care of this?

> ![](./MonteCarlo_manual.png)

Author: @allisonmcampbell reduced the image resolution so we're going to call this good for now. If there is a formal policy we want to adopt for images in the documentation we should get that settled so we can update images and use the standardized format going forward.

> ![](./MonteCarlo_manual.png)

Reviewer: I'm trying to figure that out -- I noticed that some of the plots use a lot of data points (much more than the user can actually see) so the svg version was pretty big (huge) even when compressed.

Currently looking into if [Vega-Lite](https://vega.github.io/vega-lite/) could work for just creating the charts/graphs directly from the results data in HELICS-Examples, which could enable some interesting interactive features with the visualizations in the user guide.

      "exec": "helics_broker --federates=2 --loglevel=5 --type=tcpss --port 12345",
      "host": "broker",
      "name": "broker_0",
      "loglevel": 3
    },
    {
      "directory": ".",
      "exec": "python3 Battery.py --port 12345 --seed 1",
      "host": "broker",
      "name": "Battery",
      "loglevel": 3
    },
    {
      "directory": ".",
      "exec": "python3 Charger.py --port 12345",
      "host": "broker",
      "name": "Charger",
      "loglevel": 3
    }
  ],
  "name": "Generated by make samples"
}
```

Once the samples have been created, we copy the 2 federates to the
`MERLIN_INFO` directory.

##### Study Step

We have made it to the study step. This step will execute all 10 Monte
Carlo co-simulations. There are 2 steps in the study step. The first
step is the `start_parallel_sims` step. This step will use the
helics_cli to execute each co-simulation. The second step, `cleanup`
depends on the first step. Once `start_parallel_sims` is complete the
`cleanup` step will remove any temporary data that is no longer
needed.

```yaml
study:
  - name: start_parallel_sims
    description: Run a bunch of UQ sims in parallel
    run:
      cmd: |
        spack load helics

>         /home/yee29/projects/helics/helics-cli/bin/helics run --path=$(MERLIN_INFO)/$(FED)

Author: Changed.
